libs/langchain-google-common/src/media.ts

```diff
@@ -0,0 +1,775 @@
+
+  async buildUrl(): Promise<string> {
+    return `https://generativelanguage.googleapis.com/upload/${this.apiVersion}/files`;
+  }
+}
+
+export interface AIStudioFileDownloadConnectionParams<AuthOptions>
+  extends AIStudioFileConnectionParams,
+    GoogleConnectionParams<AuthOptions> {
+  method: GoogleAbstractedClientOpsMethod;
+  name: string;
+}
+
+export class AIStudioFileDownloadConnection<
+  ResponseType extends GoogleResponse,
+  AuthOptions
+> extends GoogleDownloadConnection<AsyncCallerCallOptions, ResponseType, AuthOptions> {
+
+  method: GoogleAbstractedClientOpsMethod;
+
+  name: string;
+
+  apiVersion = "v1beta";
+
+  constructor(
+    fields: AIStudioFileDownloadConnectionParams<AuthOptions>,
+    caller: AsyncCaller,
+    client: GoogleAbstractedClient
+  ) {
+    super(fields, caller, client);
+    this.method = fields.method;
+    this.name = fields.name;
+  }
+
+  buildMethod(): GoogleAbstractedClientOpsMethod {
+    return this.method
+  }
+
+  async buildUrl(): Promise<string> {
+    return `https://generativelanguage.googleapis.com/${this.apiVersion}/files/${this.name}`;
+  }
+}
+
+export interface BlobStoreAIStudioFileBaseParams<AuthOptions>
+  extends BlobStoreGoogleParams<AuthOptions> {
+  retryTime?: number;
+}
+
+export abstract class BlobStoreAIStudioFileBase<
+  AuthOptions
+> extends BlobStoreGoogle<AIStudioFileResponse, AuthOptions> {
+
+  params?: BlobStoreAIStudioFileBaseParams<AuthOptions>;
+
+  retryTime: number = 1000;
+
+  constructor(fields?: BlobStoreAIStudioFileBaseParams<AuthOptions>) {
+    const params: BlobStoreAIStudioFileBaseParams<AuthOptions> = {
+      defaultStoreOptions: {
+        pathPrefix: "https://generativelanguage.googleapis.com/v1beta/files/",
+        actionIfInvalid: "removePath",
+      },
+      ...fields,
+    };
+    super(params);
+    this.params = params;
+    this.retryTime = params?.retryTime ?? this.retryTime ?? 1000;
+  }
+
+  _pathToName(path: string): string {
+    return path.split("/").pop() ?? path;
+  }
+
+  abstract buildAbstractedClient(
+    fields?: BlobStoreGoogleParams<AuthOptions>
+  ): GoogleAbstractedClient;
+
+  buildApiKeyClient(apiKey: string): GoogleAbstractedClient {
+    return new ApiKeyGoogleAuth(apiKey);
+  }
+
+  buildApiKey(fields?: BlobStoreGoogleParams<AuthOptions>): string | undefined {
+    return fields?.apiKey ?? getEnvironmentVariable("GOOGLE_API_KEY");
+  }
+
+  buildClient(
+    fields?: BlobStoreGoogleParams<AuthOptions>
+  ): GoogleAbstractedClient {
+    const apiKey = this.buildApiKey(fields);
+    if (apiKey) {
+      return this.buildApiKeyClient(apiKey);
+    } else {
+      // TODO: Test that you can use OAuth to access
+      return this.buildAbstractedClient(fields);
+    }
+  }
+
+  async _regetMetadata(key: string): Promise<AIStudioFileObject> {
+    // Sleep for some time period
+    // eslint-disable-next-line no-promise-executor-return
+    await new Promise(resolve => setTimeout(resolve, this.retryTime));
```

Comment:
Could we just use `AsyncCaller` logic here?

Reply:
Perhaps, but this seems like it serves a different purpose than `AsyncCaller` does and has different needs. Tho perhaps I misunderstand exactly how we can use `AsyncCaller` in this case.
* `AsyncCaller` reacts to network or HTTP errors specifically. This needs to react to the value of a returned field. (However, I guess we could change `AsyncCaller.onFailedAttempt` for this.)
* `AsyncCaller` retries immediately. The purpose here is to wait a fixed length of time before trying again. (Changing this would require changes to how we call `pRetry()`.
* `AsyncCaller` has a set number of retries. The best practice for using AI Studio File API is to wait indefinitely until the upload is either ready or has failed. (There are cases where a developer may want to upload and not care immediately if it is ready, and a negative `retryTime` takes this situation into account.

I'm open minded about using either `AsyncCaller` or `pRetry()` if you think it gains us something. I'm just not sure what we'd get in this case.